README.md
# @beardicus/line-us

A JavaScript library for controlling the [Line-us](https://line-us.com) drawing robot via its websocket interface. Works in both Node.js and the browser.

## Example

```js
const LineUs = require('@beardicus/line-us')
const bot = new LineUs()

bot.on('connected', async () => {
  // draw an 'X'
  bot.moveTo({ x: 0, y: 0 })
  bot.lineTo({ x: 500, y: 500 })
  bot.moveTo({ x: 500, y: 0 })
  bot.lineTo({ x: 0, y: 500 })

  // wait for home
  await bot.home()
  console.log(`job's done!`)
})
```

## Contents

- [Installation](#installation)
- [Setup](#setup)
- [Communication](#communication)
- [Movement](#movement)
- [Configuration](#configuration)
- [Events](#events)
- [TODO](#todo)

## Installation

Install via npm:

```shell
$ npm install @beardicus/line-us
```

A prebuilt browser version is also available via the [unpkg](https://unpkg.com/) CDN:

```html
<script type="text/javascript" src="https://unpkg.com/@beardicus/line-us/dist/line-us.min.js"></script>
```

## Setup

The code in this section is used to set up and configure a new `LineUs` instance.

### `LineUs = require('@beardicus/line-us')`

Imports the `LineUs` class. If you're using the prebuilt browser version of the library (using a `<script>` tag) you should skip this, as there will already be a global `LineUs` class available.

### `bot = new LineUs({opts})`

Create a `LineUs` instance. Optionally, pass in an object to set options:

```js
{
  url: 'ws://line-us.local',
  autoConnect: true,
  autoStart: true,
  concurrency: 3
}
```

**Parameters:** all are optional. The default values are shown above.

- **`url`**: (optional) a string that contains the websocket address to connect to. **Default is `ws://line-us.local`** which is the default name and address of all Line-us machines
- **`autoConnect`**: (optional) a boolean, if true will automatically attempt to connect to the websocket. **Defaults to `true`**
- **`autoStart`**: (optional) a boolean, if true the queue will be started automatically upon connection to the websocket. **Defaults to `true`**. See [Queue Control](#queue-control) for more information on the command queue.
- **`concurrency`**: (optional) a number which controls how many commands are sent to be buffered by the Line-us machine. This setting can improve performance where the network between the sender and Line-us has high latency. Note that higher values will create delays with the `pause()` command, as the Line-us will always execute the buffered commands before pausing. **Defaults to `3`**. This value enables relatively smooth operation on local Wifi networks, but you may want to increase it for poorer connections with more latency.

## Communication

This library uses a websocket connection to communicate with your Line-us machine. The commands in this section control the websocket connection and allow you send messages to your machine.

### `.connect()`

Connects to the Line-us machine's websocket. This is done automatically unless `autoConnect` is set to `false` when creating the `LineUs` instance.

### `.disconnect()`

Closes the websocket connection and stops the queue.

### `.send({cmd})`

Sends a command to the machine, first buffering it in the queue if necessary.

You probably don't want to use this directly, and should instead use the commands listed in the [Movement](#movement) and [Configuration](#configuration) sections to interact with the machine.

**Parameters:** accepts a single object that will be serialized into G-code. For example:

```js
{
  g: 'G01',
  params: {
    x: 1000,
    y: 900
  }
}
```

- **`g`**: (required) the G-code command
- **`params`**: (optional) an object with any parameters the G-code requires

All G-codes and parameter keys (`x` and `y` in the example above) are forced to uppercase (though the Line-us machine doesn't seem to care). The example above would be serialized into:

```
G01 X1000 Y900
```

**Returns:** all commands sent to the Line-us websocket will return a promise that resolves when we receive a response. The value will be the machine's response parsed into an object such as the following:

```js
{
  x: '350',
  y: '500',
  z: '0'
}
```

The exact data returned depends on the command being responded to.

## Movement

The commands in this section all deal with moving the robot arm. They are all asynchronous and return a promise for the Line-us machine's eventual response. See [Movement Responses](#movement-responses) for more detailed information.

You should be able to draw a masterpiece using just the canvas-style `.moveTo()` and `.lineTo()` commands, but the trio of `.penUp()` `.penDown` and plain `.to()` movements are also available for any situations where they may be more ergonomic to use.

The coordinate system used is different from the Line-us machine's default "machine coordinates". If you look at the [Line-us drawing area diagram](https://github.com/Line-us/Line-us-Programming/blob/master/Documentation/LineUsDrawingArea.pdf), this library places the `{x: 0, y: 0}` origin in the upper left of the area labeled "App drawing area". Positive numbers move to the right and down just like in the browser and the canvas API. The machine's home position in this coordinate system is `{x: 350, y: 0}`. The `z` axis behavior remains the same.

### `.moveTo({xy})`

Moves the arm to the coordinates specified, lifting the pen first if necessary.

**Parameters:** an object with one or both of `x` and `y` coordinates specified:

```js
{
  x: 800,
  y: 900
}
```

**Returns:** A promise. See [Movement Responses](#movement-responses) for details.

### `.lineTo({xy})`

Draws a line to the coordinates specified, setting the pen down first if necessary.

**Parameters:** an object with one or both of `x` and `y` coordinates specified:

```js
{
  x: 800,
  y: 900
}
```

**Returns:** A promise. See [Movement Responses](#movement-responses) for details.

### `.penUp()`

Lifts the pen up.

**Returns:** A promise. See [Movement Responses](#movement-responses) for details.

### `.penDown()`

Sets the pen down.

**Returns:** A promise. See [Movement Responses](#movement-responses) for details.

### `.to({xyz})`

Moves the arm to the coordinates specified. Use this in combination with `.penUp()` and `.penDown()` if you find `.moveTo()` and `.lineTo()` don't suit your needs.

> **Note:** You may also specify the `z` coordinate here, for manual height control, but it's not recommended. The machine will accept anything from 0–1000 for the `z` axis. This library considers any `z` coordinate < 500 to be 'down'. All other movement commands use `z: 0` and `z: 1000` exclusively when changing the pen height, so you could see unintended behaviors mixing the two strategies.

**Parameters:** an object with at least one of `x`, `y`, or `z` coordinates specified:

```js
{
  x: 800,
  y: 900,
  z: 0
}
```

**Returns:** A promise. See [Movement Responses](#movement-responses) for details.

### `.home()`

Lifts the pen up and moves to the home position at coordinates `{ x: 350, y: 0}`.

**Returns:** A promise. See [Movement Responses](#movement-responses) for details.

### `.getPosition()`

Fetches the machine's current coordinates. Because this command might end up buffered in the queue, you should instead listen for [`coordinate` events](#oncoordinates-obj) for immediate machine coordinate updates as they happen.

**Returns:** A promise. See [Movement Responses](#movement-responses) for details.

### Movement Responses

All movement commands return a promise that resolves into the Line-us machine's response. The response will be parsed into an object and will reflect the machine's new coordinates:

```js
{
  x: '350',
  y: '500',
  z: '0'
}
```

More details about the returned message object can be found in the [`.send()`](#sendcmd) command documentation.

## Queue Control

The Line-us machine originally had no buffer, and it was recommended to only send one command at a time. As of Line-us firmware version 3.0.0 this restriction has been relaxed, and it is ok to send multiple commands at once and rely on the Line-us to buffer them.

Despite this change, this library still uses a queue mechanism to optionally buffer some commands locally. The `concurrency` configuration option determines how many commands at once will be sent to the Line-us for buffering. Any commands beyond that will wait in the local queue.

The queue also provides other convenient methods of controlling the progress of your drawing. Before starting the queue you could pre-fill it with all the commands for a drawing, or you may want to start it right away and use it just as a buffer while you send commands in "real-time". You can pause, resume, and clear the queue.

### `.start()`

Starts the queue. Any messages in the queue will immediately start sending to the machine. If the queue is empty it will wait for commands.

### `.pause()`

Temporarily pause the queue. The current message(s) will finish sending and the machine will finish executing the move, so there may be a delay between pausing and the actual cessation of movement. The delay will be greater with higher values of the `concurrency` parameter.

By default if the current `z` height is less than 500, a `penUp()` command will be injected to lift the arm for the duration of the pause. The previous `z` height will be restored when the queue is resumed.

To disable the automatic lifting of the arm during pauses, pass in a configuration object with `lift: false`:

```js
bot.pause({ lift: false })
```

**Parameters:** an optional object with the following keys:

- **`lift`**: (optional) a boolean that determines if the arm should be lifted up during the pause. **Defaults to `true`**

**Returns:** a promise that resolves when the queue has flushed and the machine is actually paused.

### `.resume()`

Unpause the queue and resume sending messages where it left off. Restores the `z` axis to its previous height if the arm was lifted during the pause.

### `.stop()`

Stop the queue and clear it.

### `.clear()`

Clear all commands out of the queue. The queue remains in its current state with regards to being `paused` or `drawing`.

## Configuration

These commands get information from the machine, and set various configuration options.

### `.info`

A property of the `Line-us` instance which is populated with info provided by the machine's `hello` message:

```js
{
  version: '2.0.2Beta6 Sep 14 2018 22:17:02',
  name: 'line-us',
  serial:'999999'
}
```

Not available until the websocket is connected.

### `.getCapabilities()`

Fetches information about the machine's wifi connection mode and machine name.

**Returns:** a promise that resolves into the machine's response:

```js
{
  CONNECTMODE: 'ST',
  MACHINENAME: 'line-us'
}
```

See the official [Line-us G-code specification](https://github.com/Line-us/Line-us-Programming/blob/master/Documentation/GCodeSpec.pdf) for details on the possible values.

### `.getDiagnostics()`

Fetches extensive diagnostic information about the Line-us machine.

**Returns:** a promise that resolves into the machine's response:

```js
{
  ChipID: '990192',
  WifiMode: '1',
  WifiModeSet: '0',
  WifisConfigured: '2',
  MemDraw: '0',
  Gestures: '0',
  name: 'line-us',
  mac: '00:BE:EF:C0:FF:EE',
  FlashChipID: '1458415',
  FlashChipMode: '0',
  FlashChipSpeed: '40000000',
  FreeHeap: '26280',
  ResetReason: 'External System',
  FSUsed: '1255',
  FSTotal: '1953282',
  FSFree: '1952027',
  FSPercent: '0',
  FS: '/cal-30;/key-344; ',
  Serial: 'cd-pAd',
  Cal: '-19.70022,9.099869,-9.200562'
}
```

Currently, determining the meaning of these values is left to you, dear user, as the official [specification](https://github.com/Line-us/Line-us-Programming/blob/master/Documentation/GCodeSpec.pdf) has no further details on this command.

### `.setCalibration()`

Sets the calibration of the Line-us machine. You must manually move the arm to its calibration point before running this command. See the [Line-us help page](https://www.line-us.com/help.html#10) for more information on how to calibrate your machine.

**Returns:** a promise that resolves into the machine's response.

### `.clearCalibration()`

Clears the current calibration. You can inspect your calibration data using the [`.getDiagnostics()`](#getdiagnostics) command:

```js
response = await bot.getDiagnostics()
calibration = response.Cal
```

**Returns:** a promise that resolves into the machine's response.

### `.setMachineName('name')`

Sets the LineUs machine's name to the specified string.

**Returns:** a promise that resolves into the machine's response.

### `.getMachineName()`

Gets the LineUs machine's current name.

**Returns:** a promise that resolves into the machine's response.

### `.saveWifiNetwork({ssid, password})`

Saves a new wifi network to the machine.

**Parameters:**

- **`ssid`**: (required) the wifi network's SSID name
- **`password`**: (optional) the wifi network's password if needed

**Returns:** a promise that resolves into the machine's response.

### `.listWifiNetworks()`

Lists the wifi networks currently saved on the machine.

**Returns:** a promise that resolves into the machine's response.

### `.forgetWifiNetwork({ssid})`

Forgets the specified wifi network.

**Parameters:**

- **`ssid`**: (required) the SSID if the network to be forgotten. `*` may be used to instruct the machine to forget all saved networks.

**Returns:** a promise that resolves into the machine's response.

## Events

### `.on('state', (string) => {})`

The `state` event fires whenever there is a state change. It passes the new state as one of the following strings:

- **`disconnected`**: the websocket is not connected, either because we're just starting up or because the connection was lost or closed.
- **`connecting`**: a websocket connection is being established
- **`connected`**: the websocket is connected to the Line-us machine and has received the initial `hello` message.
- **`paused`**: the queue is paused and no messages will be sent to the machine.
- **`drawing`**: the queue is started and messages are being sent to the machine.

> **Note:** all the states emitted under the `state` event are also emitted under their own name. So if you want to listen only for the `connected` event, for example, you may do `.on('connected', () => {})`.

### `.on('error', (string) => {})`

Emits errors from the websocket or Line-us machine.

#### `.on('coordinates', (obj) => {})`

Emits a coordinate object every time the Line-us machine replies with its current coordinates. The object will be in the following form:

```js
{
  x: 800,
  y: 150,
  z: 1000
}
```

It will always have all three coordinates, even if the command that triggered the event provided only one or two. The coordinates are emitted _when the machine finishes executing the current command_.

## TODO

- [x] Better error handling
- [x] Validate inputs
- [x] Tests
- [ ] Functional tests
- [x] Fix browser tests
- [ ] Emit simple progress updates using queue size / total
- [ ] Emit smarter progress updates with time estimates


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [2.0.0] 2019-06-27

### Added

- A `concurrency` parameter has been added to the constructor, with a default value of `3`. This parameter controls how many commands can be "in flight" and buffered by the Line-us machine. Setting this to higher values can smooth out operation over unreliable or high-latency connections. The default value of `3` seems good for most local Wifi situations.

### Fixed

- `.coordinates` will now be set to the home coordinates (`{x: 350, y: 0, z: 1000}`) after a connection is established, as the Line-us machine does a homing sequence as part of the connection process. Previously we would return `undefined` coordinates until after the first move.

## [2.0.0] 2019-01-29

### Changed

- `.pause()` now lifts the arm by default, and returns it to the previous `z` height upon `.resume()`. It returns a promise that resolves when the machine is actually paused.

### Fixed

- `npm run test-browser` now finishes and exits properly.
- The README erroneously stated that coordinates were emitted from the machine as soon as a command is accepted at the beginning of a movement. The response actually comes when the movement is complete. Updated README to reflect that `coordinate` events are emitting the coords of the _last completed movement_.

## [1.0.0] 2019-01-20

### Added

- Everything.


lib/LineUs.js
const WebSocket = require('isomorphic-ws')
const pQueue = require('p-queue')
const pEvent = require('p-event')
const Nanobus = require('nanobus')
const lowercaseKeys = require('lowercase-keys')

const { serialize, deserialize } = require('./serialize.js')
const { translate, untranslate } = require('./translate.js')
const validate = require('./validate.js')

class LineUs extends Nanobus {
  constructor(opts) {
    opts = Object.assign(
      {
        url: 'ws://line-us.local',
        autoConnect: true,
        autoStart: true,
        concurrency: 3,
      },
      opts
    )
    super('line-us')

    this.url = opts.url
    this.autoConnect = opts.autoConnect
    this.autoStart = opts.autoStart
    this.concurrency = opts.concurrency

    this.info = {} // populated by `hello` message from Line-us

    this._ws = null
    this._queue = new pQueue({
      concurrency: this.concurrency,
      autoStart: false,
    })
    this._responseQueue = []

    this._state = 'disconnected'
    this._coordinates = {}
    this._queuedPenState = 'up'

    if (this.autoConnect) {
      this.connect()
    }
  }

  get state() {
    return this._state
  }
  set state(state) {
    this._state = state
    this.emit('state', state)
    this.emit(state)
  }

  get coordinates() {
    return this._coordinates
  }
  set coordinates(xyz) {
    this._coordinates = xyz
    this.emit('coordinates', xyz)
  }

  connect() {
    this.state = 'connecting'
    this._ws = new WebSocket(this.url)

    this._ws.onopen = async () => {
      let response = await pEvent(this._ws, 'message') // wait for hello
      response = deserialize(response.data || response)

      this.info = lowercaseKeys(response.data)
      this.state = 'connected'
      this._queuedPenState = 'up' // should be at home position right now
      this.coordinates = { x: 350, y: 0, z: 1000 }

      this._ws.onmessage = (response) => {
        // resolve promises with responses as they come in
        this._responseQueue.shift()(response)
      }

      if (this.autoStart) {
        this.start()
      } else {
        this.pause() // we're already paused, but this will emit the event
      }
    }

    this._ws.onclose = () => {
      this.state = 'disconnected'
      this.stop()
    }

    this._ws.onerror = (e) => {
      this.emit('error', 'LineUs: websocket error: ' + e.error.message)
    }
  }

  disconnect() {
    // it's possible this.stop() should go here instead of waiting for `onclose`
    this._ws.close()
  }

  start() {
    this._queue.start()
    this.state = 'drawing'
  }

  async pause({ lift = true } = {}) {
    if (this.state === 'paused') return Promise.resolve() // resolve immediately
    validate.lift(lift)

    // we add the pause command at the head of the queue w/ `priority: 1`
    // this allows the currently executing command to flush, and lets us
    // test the current z height to determine if we should send a `penUp()`
    // and queue the corresponding `penDown()` for execution when we `resume()`
    return this._queue.add(
      () => {
        const z = this.coordinates.z

        this._queue.pause()
        this.state = 'paused'

        if (z < 500 && lift) {
          this._queue.add(
            () => {
              return this._send({ g: 'G01', params: { z } })
            },
            { priority: 1 }
          )
          return this._send({ g: 'G01', params: { z: 1000 } })
        }
      },
      { priority: 1 }
    )
  }

  resume() {
    this.start()
  }

  async stop() {
    await this.pause()
    this.clear()
  }

  clear() {
    this._queue.clear()
  }

  async send(cmd) {
    validate.cmd(cmd)
    const { g, params } = cmd

    if (params && 'z' in params) {
      // monitor z to keep track of the pen up/down state of queued commands
      if (params.z < 500) {
        this._queuedPenState = 'down'
      } else {
        this._queuedPenState = 'up'
      }
    }

    if (g.toUpperCase() === 'G28') {
      // special case pen up state for 'home' command
      this._queuedPenState = 'up'
    }

    // queue up calls to _send because Line-us has no buffer
    return this._queue.add(() => this._send(cmd))
  }

  async _send(cmd) {
    let responseEvent = new Promise((resolve, reject) => {
      this._responseQueue.push(resolve)
      // TODO: some sort of timeout-based rejection here?
    })

    this._ws.send(serialize(cmd))

    let response = await responseEvent
    response = deserialize(response.data || response)

    switch (response.type) {
      case 'ok':
        let data = response.data
        if ('X' in data) {
          data = untranslate(lowercaseKeys(data)) // translate back from machine coordinates
          this.coordinates = data // update and emit current coordinates
        }
        return data

      case 'error':
        throw new Error('LineUs: ' + response.data.INFO)

      default:
        throw new Error('LineUs: unknown response type')
    }
  }

  async penUp() {
    return this.to({ z: 1000 })
  }

  async penDown() {
    return this.to({ z: 0 })
  }

  async to(xyz) {
    validate.xyz(xyz)
    return this.send({
      g: 'G01',
      params: translate(xyz),
    })
  }

  async moveTo(xy) {
    validate.xy(xy)
    // to lift the pen, we inject Z1000 into the move
    if (this._queuedPenState === 'down') xy.z = 1000
    return this.to(xy)
  }

  async lineTo(xy) {
    validate.xy(xy)
    // to lower the pen, send a separate Z0 command first
    // combining z w/ the xy move does not work because for a move with {z: 0}
    // Line-us won't drop the pen until the end of the movement
    if (this._queuedPenState === 'up') this.penDown() // TODO: don't ignore this response?
    return this.to(xy)
  }

  async home() {
    return this.send({ g: 'G28' })
  }

  async getPosition() {
    return this.send({ g: 'M114' })
  }

  async getCapabilities() {
    return this.send({ g: 'M115' })
  }

  async getDiagnostics() {
    return this.send({ g: 'M122' })
  }

  async setCalibration() {
    return this.send({ g: 'M374' })
  }

  async clearCalibration() {
    return this.send({
      g: 'M374',
      params: { s: 'clear' },
    })
  }

  async setMachineName(name) {
    validate.name(name)
    return this.send({
      g: 'M550',
      params: { p: `"${name}"` }, // add quotes for safety
    })
  }

  async getMachineName() {
    // TODO: extract MACHINENAME from response and return just the string?
    return this.send({ g: 'M550' })
  }

  async saveWifiNetwork(input) {
    validate.wifi(input)
    const { ssid, password } = input
    let params = { s: `"${ssid}"` }

    if (password !== undefined) params.p = `"${password}"`

    return this.send({
      g: 'M587',
      params: params,
    })
  }

  async listWifiNetworks() {
    return this.send({ g: 'M587' })
  }

  async forgetWifiNetwork(input) {
    validate.wifi(input)
    const { ssid } = input
    const params = { s: `"${ssid}"` }

    return this.send({
      g: 'M588',
      params: params,
    })
  }
}

module.exports = LineUs
